### PR TITLE
boost: conan v2 support

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [ ] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
FYI, there are some tricks to avoid to rely on conan v1 helpers:
https://github.com/conan-io/conan/issues/12603
https://github.com/conan-io/conan/issues/12604
https://github.com/conan-io/conan/issues/12605

File: recipes/boost/all/conanfile.py [hunk 1]
```diff
@@ -865,18 +835,16 @@ def build(self):
...
-        use_vcvars = self._is_msvc and not self.settings.compiler.get_safe("toolset", default="")
-        with tools.vcvars(self.settings) if use_vcvars else tools.no_op():
-            with chdir(self, sources):
-                # To show the libraries *1
-                # self.run("%s --show-libraries" % b2_exe)
-                self.run(full_command, run_environment=True)
+        with chdir(self, sources):
+            # To show the libraries *1
+            # self.run("%s --show-libraries" % b2_exe)
+            self.run(full_command)
```
Comment (SpaceIm) [inline recipes/boost/all/conanfile.py:847]: might break now: https://github.com/conan-io/conan/issues/12606

**Review by jwillikers — DISMISSED** [review]
Thanks for fixing the order of all the Boost versions. Will the B2 generator be ported to Conan V2? It would be nice to be able to use that for the Boost recipe... assuming it would be possible.

One of these days we should probably fix the missing system libraries warning.

**Comment by prince-chrismc** [comment]
Given how important this recipe is we need to take our time testing, given the backlog anyone willing should check out this PR and share their results in 1.x or 2.0 and let us know ❤️

**Comment by SpaceIm** [comment]
> Given how important this recipe is we need to take our time testing, given the backlog anyone willing should check out this PR and share their results in 1.x or 2.0 and let us know ❤️

Unless you ping some people, I don't think you can hope any feedback from users until it's merged.

**Comment by jwillikers** [comment]
@prince-chrismc @SpaceIm I've tested building and using this fork of the Boost Conan package and it works for me without issue.

**Comment by prince-chrismc** [comment]
So 2.0 does not work 

```log
ERROR: Package 'libiconv/1.17' not resolved: libiconv/1.17: Cannot load recipe.
Error loading conanfile at '/Users/christopherm/.conan2/p/41879da388a50bdc/e/conanfile.py': Unable to load conanfile in /Users/christopherm/.conan2/p/41879da388a50bdc/e/conanfile.py
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/christopherm/.conan2/p/41879da388a50bdc/e/conanfile.py", line 18, in <module>
    from conans.tools import get_gnu_triplet
ModuleNotFoundError: No module named 'conans.tools'
```

File: recipes/boost/all/conanfile.py [hunk 2]
```diff
@@ -442,55 +451,18 @@ def validate(self):
...
-                    min_compiler_version = self._min_compiler_version_default_cxx11
-                    if min_compiler_version is not None:
-                        if Version(self.settings.compiler.version) < min_compiler_version:
-                            raise ConanInvalidConfiguration("Boost.Wave requires (boost:)cppstd>=11 (current one is lower)")
+            version_cxx11_standard = self._min_compiler_version_default_cxx11
+            if version_cxx11_standard and Version(self.settings.compiler.version) < version_cxx11_standard:
+                raise ConanInvalidConfiguration(
+                    f"Boost.{{{','.join(self._cxx11_boost_libraries)}}} requires a c++11 compiler "
+                    "(please set compiler.cppstd or use a newer compiler)"
+                )
```
Comment (prince-chrismc) [inline recipes/boost/all/conanfile.py:465]: This section does not like me 

```log
boost/1.80.0: Exported revision: df465c37ceb1d7766e55d739265db246
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.cppstd=17
compiler.libcxx=libc++
compiler.version=13
os=Macos
os_build=Macos
[options]
[build_requires]
[env]

bzip2/1.0.8: Main binary package '5eea6cd0d4094bfb10578864f66812e7c6a4d0e6' missing. Using compatible package 'f2848f95be36a5d770fc8e833993b621f703b507'
libbacktrace/cci.20210118: Main binary package '299aaabcb5fb1385c096ccf2c550a653c4438173' missing. Using compatible package 'e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc'
libiconv/1.17: Main binary package '299aaabcb5fb1385c096ccf2c550a653c4438173' missing. Using compatible package 'e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc'
zlib/1.2.13: Main binary package '299aaabcb5fb1385c096ccf2c550a653c4438173' missing. Using compatible package 'e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc'
boost/1.80.0 (test package): Installing package
Requirements
    boost/1.80.0 from local cache - Cache
    bzip2/1.0.8 from 'conancenter' - Cache
    libbacktrace/cci.20210118 from 'conancenter' - Cache
    libiconv/1.17 from 'conancenter' - Cache
    zlib/1.2.13 from 'conancenter' - Cache
Packages
    boost/1.80.0:INVALID - Invalid
    bzip2/1.0.8:f2848f95be36a5d770fc8e833993b621f703b507 - Cache
    libbacktrace/cci.20210118:e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc - Cache
    libiconv/1.17:e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc - Cache
    zlib/1.2.13:e4fbacdee83eb1a2e68227ffe19f6e9a61b36cbc - Cache

Installing (downloading, building) binaries...
ERROR: There are invalid packages (packages that cannot exist for this configuration):
boost/1.80.0: Invalid ID: Boost.{fiber,json,math,nowide,wave} requires a c++11 compiler (please set compiler.cppstd or use a newer compiler)
```

Comment (prince-chrismc) [reply recipes/boost/all/conanfile.py:465]: The old code had an extra else https://github.com/conan-io/conan-center-index/pull/14415/files#diff-cc14160728d1ad321c1444ffad82615139181e681f4acb7dfd045a005cc8bd66L464, this should nto run with cppstd set

File: recipes/boost/all/test_package/conanfile.py [hunk 3]
```diff
@@ -1,94 +1,101 @@
...
+    def layout(self):
+        cmake_layout(self)
+
+    def requirements(self):
+        self.requires(self.tested_reference_str)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["HEADER_ONLY"] = self.dependencies["boost"].options.header_only
+        if not self.dependencies["boost"].options.header_only:
```
Comment (prince-chrismc) [inline recipes/boost/all/test_package/conanfile.py:30]: This option does not like me either

```log
boost/1.80.0 (test package): Calling build()
boost/1.80.0 (test package): CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/Users/christopherm/conan-center-index/recipes/boost/all/test_package/build/generators/conan_toolchain.cmake" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/Users/christopherm/conan-center-index/recipes/boost/all/test_package/."
-- Using Conan toolchain: /Users/christopherm/conan-center-index/recipes/boost/all/test_package/build/generators/conan_toolchain.cmake
-- Conan: Component target declared 'Boost::headers'
-- Conan: Component target declared 'Boost::boost'
-- Conan: Target declared 'boost::boost'
-- Configuring done
CMake Error at CMakeLists.txt:12 (target_link_libraries):
  Target "random_exe" links to:

    Boost::random

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:18 (target_link_libraries):
  Target "regex_exe" links to:

    Boost::regex

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:24 (target_link_libraries):
  Target "test_exe" links to:

    Boost::unit_test_framework

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:30 (target_link_libraries):
  Target "coroutine_exe" links to:

    Boost::coroutine

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:36 (target_link_libraries):
  Target "chrono_exe" links to:

    Boost::chrono

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:63 (target_link_libraries):
  Target "locale_exe" links to:

    Boost::locale

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:70 (target_link_libraries):
  Target "stacktrace_addr2line_exe" links to:

    Boost::stacktrace_addr2line

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:76 (target_link_libraries):
  Target "stacktrace_backtrace_exe" links to:

    Boost::stacktrace_backtrace

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:84 (target_link_libraries):
  Target "stacktrace_noop_exe" links to:

    Boost::stacktrace_noop

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



CMake Error at CMakeLists.txt:97 (target_link_libraries):
  Target "stacktrace_basic_exe" links to:

    Boost::stacktrace_basic

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
ERROR: boost/1.80.0 (test package): Error in build() method, line 55
	cmake.configure()
	ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/Users/christopherm/conan-center-index/recipes/boost/all/test_package/build/generators/conan_toolchain.cmake" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/Users/christopherm/conan-center-index/recipes/boost/all/test_package/."
 ✘ christopherm@christopherm-mac  ~/conan-center-index/recipes/boost   boost-conan-v2  conan create all boost/1.80.0@ -o boost:header_only=True
```

File: recipes/boost/all/test_package/conanfile.py [hunk 4]
```diff
@@ -1,94 +1,101 @@
...
+            tc.variables["Boost_USE_STATIC_LIBS"] = not self.dependencies["boost"].options.shared
+        tc.variables["WITH_PYTHON"] = not self.dependencies["boost"].options.without_python
+        if not self.dependencies["boost"].options.without_python:
+            pyversion = Version(self.dependencies["boost"].options.python_version)
+            tc.variables["Python_ADDITIONAL_VERSIONS"] = f"{pyversion.major}.{pyversion.minor}"
+            tc.variables["PYTHON_COMPONENT_SUFFIX"] = f"{pyversion.major}.{pyversion.minor}"
+        tc.variables["WITH_RANDOM"] = not self.dependencies["boost"].options.without_random
+        tc.variables["WITH_REGEX"] = not self.dependencies["boost"].options.without_regex
+        tc.variables["WITH_TEST"] = not self.dependencies["boost"].options.without_test
+        tc.variables["WITH_COROUTINE"] = not self.dependencies["boost"].options.without_coroutine
```
Comment (prince-chrismc) [inline recipes/boost/all/test_package/conanfile.py:40]: I tried testing this one and it does not like me either

```log
boost/1.80.0 (test package): Calling build()
boost/1.80.0 (test package): CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/Users/christopherm/conan-center-index/recipes/boost/all/test_package/build/generators/conan_toolchain.cmake" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/Users/christopherm/conan-center-index/recipes/boost/all/test_package/."
-- Using Conan toolchain: /Users/christopherm/conan-center-index/recipes/boost/all/test_package/build/generators/conan_toolchain.cmake
-- Conan: Component target declared 'Boost::diagnostic_definitions'
-- Conan: Component target declared 'Boost::disable_autolinking'
-- Conan: Component target declared 'Boost::dynamic_linking'
-- Conan: Component target declared 'Boost::headers'
-- Conan: Component target declared 'Boost::boost'
-- Conan: Component target declared 'boost::_libboost'
-- Conan: Component target declared 'Boost::atomic'
-- Conan: Component target declared 'Boost::container'
-- Conan: Component target declared 'Boost::context'
-- Conan: Component target declared 'Boost::date_time'
-- Conan: Component target declared 'Boost::exception'
-- Conan: Component target declared 'Boost::program_options'
-- Conan: Component target declared 'Boost::regex'
-- Conan: Component target declared 'Boost::serialization'
-- Conan: Component target declared 'Boost::stacktrace'
-- Conan: Component target declared 'Boost::stacktrace_addr2line'
-- Conan: Component target declared 'Boost::stacktrace_backtrace'
-- Conan: Component target declared 'Boost::stacktrace_basic'
-- Conan: Component target declared 'Boost::stacktrace_noop'
-- Conan: Component target declared 'Boost::system'
-- Conan: Component target declared 'Boost::test'
-- Conan: Component target declared 'Boost::test_exec_monitor'
-- Conan: Component target declared 'Boost::wserialization'
-- Conan: Component target declared 'Boost::chrono'
-- Conan: Component target declared 'Boost::filesystem'
-- Conan: Component target declared 'Boost::prg_exec_monitor'
-- Conan: Component target declared 'Boost::random'
-- Conan: Component target declared 'Boost::thread'
-- Conan: Component target declared 'Boost::timer'
-- Conan: Component target declared 'Boost::type_erasure'
-- Conan: Component target declared 'Boost::unit_test_framework'
-- Conan: Component target declared 'Boost::contract'
-- Conan: Component target declared 'Boost::iostreams'
-- Conan: Component target declared 'Boost::locale'
-- Conan: Component target declared 'Boost::log'
-- Conan: Component target declared 'Boost::log_setup'
-- Conan: Target declared 'boost::boost'
-- Conan: Target declared 'BZip2::BZip2'
-- Conan: Including build module from '/Users/christopherm/.conan/data/bzip2/1.0.8/_/_/package/f2848f95be36a5d770fc8e833993b621f703b507/lib/cmake/conan-official-bzip2-variables.cmake'
-- Conan: Target declared 'ZLIB::ZLIB'
-- Conan: Target declared 'Iconv::Iconv'
-- Conan: Target declared 'libbacktrace::libbacktrace'
-- Configuring done
CMake Error at CMakeLists.txt:30 (target_link_libraries):
  Target "coroutine_exe" links to:

    Boost::coroutine

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
ERROR: boost/1.80.0 (test package): Error in build() method, line 55
	cmake.configure()
	ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/Users/christopherm/conan-center-index/recipes/boost/all/test_package/build/generators/conan_toolchain.cmake" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/Users/christopherm/conan-center-index/recipes/boost/all/test_package/."
 ✘ christopherm@christopherm-mac  ~/conan-center-index/recipes/boost   boost-conan-v2  conan create all boost/1.80.0@  -o boost:without_coroutine=True
```

**Review by prince-chrismc — COMMENTED** [review]
I ran a few tests on my Mac book pro and had a few issues, the default one works but maybe you could try the ones that failed for me and let me know if it's on my end but seems theres a few edge cases

I'll try add test this one my windows + linux before I leave on vacation tomorrow if dear future me here's the commands I tested

```sh
 5871  conan create all --version 1.80.0
 5872  pip install conan==1.53
 5873  conan create all boost/1.80.0@
 5874  conan create all boost/1.80.0@ -s compiler.cppstd=17
 5875  conan create all boost/1.80.0@ -o boost:header_only=True
 5876  conan create all boost/1.80.0@ -o boost:without_regex=True
 5877  conan create all boost/1.80.0@ -o boost:without_regex=True -o boost:without_iostreams
 5878  conan create all boost/1.80.0@ -o boost:without_regex=True -o boost:without_iostreams=True
 5879  conan create all boost/1.80.0@  -o boost:without_iostreams=True
 5880  conan create all boost/1.80.0@  -o boost:without_coroutine=True`
```

**Review by jwillikers — APPROVED** [review]

File: recipes/boost/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/boost/all/conanfile.py]: Indeed, I've restored the `else`.

File: recipes/boost/all/test_package/conanfile.py (hunk 3 above)
Comment (SpaceIm) [reply recipes/boost/all/test_package/conanfile.py:30]: `boost:header_only` option is supposed to be tested in c3i, so I don't understand how it can fail on your machine.

Comment (SpaceIm) [reply recipes/boost/all/test_package/conanfile.py:30]: Maybe you have not cleared build folder and CMakeUserPresets of test package after a conan create with another set of options, it's a common mistake when you test conan v2 generators in test package.

File: recipes/boost/all/test_package/conanfile.py (hunk 4 above)
Comment (SpaceIm) [reply recipes/boost/all/test_package/conanfile.py:40]: I can't reproduce this error on my mac, I guess you have forgotten to clear build folder of test package.

**Comment by SpaceIm** [comment]
> I ran a few tests on my Mac book pro and had a few issues, the default one works but maybe you could try the ones that failed for me and let me know if it's on my end but seems theres a few edge cases
> 
> I'll try add test this one my windows + linux before I leave on vacation tomorrow if dear future me here's the commands I tested
> 
> ```shell
>  5871  conan create all --version 1.80.0
>  5872  pip install conan==1.53
>  5873  conan create all boost/1.80.0@
>  5874  conan create all boost/1.80.0@ -s compiler.cppstd=17
>  5875  conan create all boost/1.80.0@ -o boost:header_only=True
>  5876  conan create all boost/1.80.0@ -o boost:without_regex=True
>  5877  conan create all boost/1.80.0@ -o boost:without_regex=True -o boost:without_iostreams
>  5878  conan create all boost/1.80.0@ -o boost:without_regex=True -o boost:without_iostreams=True
>  5879  conan create all boost/1.80.0@  -o boost:without_iostreams=True
>  5880  conan create all boost/1.80.0@  -o boost:without_coroutine=True`
> ```

Except `conan create all boost/1.80.0@ -s compiler.cppstd=17` which has been fixed in latest commit for compilers whose default C++ standard was lower than C++11, I think all other issues you have seen come from the fact that your build folder in test package was not cleared before running again `conan create` (quite tedious side effects of conan v2 generators in test package).

**Comment by prince-chrismc** [comment]
I did not clear the test build folder 😵‍💫 yeah v2 does that for free now I didn't realize.

 Cheers, on my calendar I will circle back since this is important :)

**Review by jwillikers — APPROVED** [review]

**Comment by prince-chrismc** [comment]
I tested these combos on my mac and the results are as expected... Going to try this out on my windows this evening

```
5888  rm -rf all/test_package/build && rm all/test_package/CMakeUserPresets.json
 5889  conan create all boost/1.80.0@ -s compiler.cppstd=17
 5890  rm -rf all/test_package/build && rm all/test_package/CMakeUserPresets.json
 5891  conan create all boost/1.80.0@ -o boost:without_regex=True -o boost:without_iostreams=True
 5892  conan create all boost/1.80.0@ -o boost:without_regex=False -o boost:without_iostreams=True
 5893  rm -rf all/test_package/build && rm all/test_package/CMakeUserPresets.json
 5894  conan create all boost/1.80.0@  -o boost:without_coroutine=True
 5895  conan create all boost/1.80.0@  -o boost:without_coroutine=True -s compiler.cppstd=14
 5896  rm -rf all/test_package/build && rm all/test_package/CMakeUserPresets.json
 5897  conan create all boost/1.80.0@  -o boost:without_coroutine=True -s compiler.cppstd=14
 5899  conan create all boost/1.80.0@  -o boost:without_math=True -s compiler.cppstd=14
 5900  conan create all boost/1.80.0@  -o boost:without_math=False -s compiler.cppstd=98
```

**Comment by prince-chrismc** [comment]
Windows testing

```
 16 conan-script create all boost/1.80.0@
  20  rm all/test_package/build
  21  rm all/test_package/CMakeUserPresets.json
  22 conan-script create all boost/1.80.0@ -s compiler.cppstd=17
  23  rm all/test_package/CMakeUserPresets.json
  30  rm all/test_package/build
  31 conan-script create all boost/1.80.0@ -o boost:without_regex=True -o boost:without_iostreams=True
  32 conan-script create all boost/1.80.0@ -o boost:without_regex=False -o boost:without_iostreams=True
  33  rm all/test_package/build
  34  rm all/test_package/CMakeUserPresets.json
  35 conan-script create all boost/1.80.0@ -o boost:without_coroutine=True -s compiler.cppstd=98
  36 conan-script create all boost/1.80.0@ -o boost:without_coroutine=True -s compiler.cppstd=14
  39 conan-script create all boost/1.80.0@ -o boost:without_fiber=True
```

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
I did the best I can testing this... if there are any breaks please help us fix them and document how to test them please!

**Comment by prince-chrismc** [comment]
Hopefully!

**Comment by jsaf0** [comment]
This change seems to break cross compilation of Boost with host profiles that uses CXX and CC variables in the [env] section of the profile (just like the documentation describes).

**Comment by SpaceIm** [comment]
@jsaf0 it has already been mentioned here: https://github.com/conan-io/conan-center-index/issues/14767. You can open an issue in conan client side to ask for backward compatibility of `[env]` in new conan helpers, but I'm not sure it will be accepted.

**Comment by prince-chrismc** [comment]
The `[env]` is deprecated you should use `buildenv` and `runenv` :) hopefully that tip helps you out

**Comment by SpaceIm** [comment]
https://github.com/conan-io/conan-center-index/issues/14781
